### PR TITLE
Make parallel tests insensitive to user configuration

### PR DESCRIPTION
**Pull request by astrofrog**
This is an attempt to fix https://github.com/astropy/astropy/issues/8691 - I discovered that when run in parallel mode, the user configuration is **not** ignored.

The first step here is to add a regression test by writing out to the config file before running the parallel tests on CircleCI. This failed, so I'm now pushing some fixes.

It turns out there are several places where we need to set the config/cache to be temporary directories, not just in the test runner:

* ``astropy/conftest.py`` - we need to set the path to the temporary config/cache directories here for cases when running the tests in parallel mode using e.g.

```
python setup.py test --parallel
```

This is because pytest starts up new subprocesses, which aren't aware of the temporary directories set in the test runner.

* ``conftest.py`` - the temporary paths need to be set here for when pytest is run directly on an astropy package, e.g.

```
pytest astropy/io/votable
```

* ``docs/conftest.py`` - we also need to set them here for when running the tests with ``python setup.py test -t docs/...`` since then the root-level ``conftest.py`` gets ignored (since the tests run from outside the source directory, and ``astropy/conftest.py`` gets ignored since we are only testing things in ``docs``

This will become simpler some day if we switch to only using ``pytest`` directly.

Fixes https://github.com/astropy/astropy/issues/8154

### CODE REVIEW

**Comment by astrofrog** [comment]
@saimn - could you check if this fixes #8691 for you?

**Comment by astrofrog** [comment]
The Matplotlib developer build can be ignored, this is due I think to the 3.1.x release today.

**Comment by saimn** [comment]
Sadly this doesn't fix #8691, but I agree that this looks a good idea.

**Comment by astrofrog** [comment]
@saimn - huh weird it does fix it for me locally... (and I could reproduce your issue). Just to be clear, you still get the test failure with:

```
python setup.py test -R -j 2 -t docs/timeseries/index.rst
```

? or are you running the tests differently?

**Comment by saimn** [comment]
@astrofrog - Yes, this command and no user configuration. How did you manage to reproduce?

**Comment by astrofrog** [comment]
@saimn - well I had to set the user configuration to include ``max_width`` and then run in parallel mode (which didn't ignore the user config). And this gave the behavior of working in serial mode and not parallel mode. Are you seeing the issue only in parallel mode?

**Comment by saimn** [comment]
@astrofrog - ok, I agree that there is an potential issue with local config in parallel mode, which is good to fix, but in my case this is not the problem. In my case it seems that the COLUMNS environment variable is set in parallel mode (https://github.com/astropy/astropy/issues/8691#issuecomment-493710823), which makes `terminal_size` to return the real terminal size. I don't yet understand where this variable comes from.

**Comment by astrofrog** [comment]
@saimn - ah ok I see. Regardless of why xdist does that, we should probably make sure we set COLUMNS and LINES to constant values in the ``conftest.py`` files since it seems that's something we'd want to ignore, especially if a user had those set by default?

**Comment by saimn** [comment]
Reading a bit more (https://unix.stackexchange.com/questions/215584/whats-the-name-of-the-environment-variable-with-current-terminal-width), COLUMNS (and LINES) is only a shell variable and not exported into the environment. So they are not supposed to be set manually, the question is why in this specific case it is available as an environment variable.

**Comment by astrofrog** [comment]
Thanks for looking into this further! So I think the way forward is to still consider this PR since it addresses https://github.com/astropy/astropy/issues/8154, and then to not consider #8691 as critical for v3.2 since it's a reasonably advanced use case and doesn't seem to happen systematically on all machines. Let's continue to discuss that and explore possibilities in #8691.

File: astropy/conftest.py [hunk 1]
```diff
@@ -52,6 +65,9 @@ def pytest_unconfigure(config):
         matplotlib.rcParams.update(matplotlibrc_cache)
         matplotlibrc_cache.clear()
 
+    os.environ.pop('XDG_CONFIG_HOME')
+    os.environ.pop('XDG_CACHE_HOME')
+
```
Comment (saimn) [inline astropy/conftest.py]: Why is this needed here but not in the other conftest files ?

**Comment by saimn** [comment]
Looks good to me, just a small question. Btw when running the tests a few days ago with pytest I found my user cache was deleted by the `test_download_cache` test (https://github.com/saimn/astropy/commit/b5a4da590d661a85ea35e281e49a2cf299e35aeb) :), so I'm happy to see a better fix. 

It may be a bit sad to have to download all test files each time though. Could we instead put the test files in a specify directory in `~/.astropy` ?

File: astropy/conftest.py (hunk 1 above)
Comment (pllim) [reply astropy/conftest.py]: I don't know how I feel about playing with env var in general in `conftest`. Let's say user sets custom vars, then run tests, `conftest` messes them up, now user complains that the these vars are not what they set anymore. Can this happen?

Comment (astrofrog) [reply astropy/conftest.py]: The reason it's needed here and not in the other ``conftest.py`` files is because this is the only one which might be used inside a persistent python session such as IPython. The other ones will only matter if you run ``pytest``, and the Python process will end at the end of the tests. Any changes to ``os.environ`` don't change the variables beyond the Python process:

```
Python 3.7.1 (v3.7.1:260ec2c36a, Oct 20 2018, 03:13:28) 
[Clang 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.environ['FOO'] = 'BAR'
>>> 
$ echo $FOO
$
```

However, in this specific ``conftest.py`` file I should probably indeed save any pre-existing variables and reset then in ``unconfigure``.

**Comment by astrofrog** [comment]
@saimn @pllim - I've now added code to restore the variables if the user is testing from e.g. IPython. I've also added a note to the other ``conftest.py`` files to explain why we don't need to unset the variables.

Good question about avoiding downloading all the files every time. Strictly speaking it's probably cleaner since it ensures an isolated environment, but if we wanted to avoid that we could change just the config but not the cache path for the tests.

**Comment by pllim** [comment]
> It would be good to find a way to avoid downloading all the files every time

@saimn , feel free to open a new issue about it lest we forget. Maybe the problem would solve itself if we switch to `git lfs`... ?
